contrib/format-msgpack/src/main/java/org/apache/drill/exec/store/msgpack/MsgpackSchema.java

```diff
@@ -0,0 +1,114 @@
+    MaterializedField previousMapField = null;
+    if (schemaLocation != null && fileSystem.exists(schemaLocation)) {
+      try (FSDataInputStream in = fileSystem.open(schemaLocation)) {
+        String schemaData = IOUtils.toString(in);
+        Builder newBuilder = SerializedField.newBuilder();
+        try {
+          TextFormat.merge(schemaData, newBuilder);
+        } catch (ParseException e) {
+          throw new DrillRuntimeException("Failed to read schema file: " + schemaLocation, e);
+        }
+        SerializedField read = newBuilder.build();
+        previousMapField = MaterializedField.create(read);
+      }
+    }
+    return previousMapField;
+  }
+
+  public void save(MaterializedField mapField, Path schemaLocation) throws IOException {
+    try (FSDataOutputStream out = fileSystem.create(schemaLocation, true)) {
+      SerializedField serializedMapField = mapField.getSerializedField();
+      String data = TextFormat.printToString(serializedMapField);
+      IOUtils.write(data, out);
+    }
+  }
+
+  public MaterializedField merge(MaterializedField existingField, MaterializedField newField) {
+    Preconditions.checkArgument(existingField.getType().getMinorType() == MinorType.MAP,
+        "Field " + existingField + " is not a MAP type.");
+    Preconditions.checkArgument(existingField.hasSameTypeAndMode(newField),
+        "Field " + existingField + " and " + newField + " not same.");
+    MaterializedField merged = existingField.clone();
+    privateMerge(merged, newField);
+    return merged;
+  }
+
+  private void privateMerge(MaterializedField existingField, MaterializedField newField) {
+    Preconditions.checkArgument(existingField.getType().getMinorType() == MinorType.MAP,
+        "Field " + existingField + " is not a MAP type.");
+    for (MaterializedField newChild : newField.getChildren()) {
+      String newChildName = newChild.getName();
+      MaterializedField foundExistingChild = getFieldByName(newChildName, existingField);
+      if (foundExistingChild != null) {
+        if (foundExistingChild.hasSameTypeAndMode(newChild)) {
+          if (foundExistingChild.getType().getMinorType() == MinorType.MAP) {
+            privateMerge(foundExistingChild, newChild);
+          } // else we already have it
+        } else {
+          // error
+          throw new SchemaChangeRuntimeException("Not the same schema for " + foundExistingChild + " and " + newChild);
+        }
+      } else {
+        existingField.addChild(newChild.clone());
+      }
+    }
+  }
+
+  private MaterializedField getFieldByName(String newChildName, MaterializedField existingField) {
+    for (MaterializedField f : existingField.getChildren()) {
+      if (newChildName.equalsIgnoreCase(f.getName())) {
+        return f;
+      }
+    }
+    return null;
+  }
+
+  public Path findSchemaFile(Path dir) throws IOException {
+    int MAX_DEPTH = 5;
+    int depth = 0;
+    while (dir != null && depth < MAX_DEPTH) {

```

Comment:
Should this be done once in the planner, and passed in the plan to each reader? Otherwise, doing this search for every file might be a bit slow.

Reply:
yes, I've already changed that. Thanks.